dali/pipeline/operators/arithmetic/arithmetic.h

```diff
@@ -0,0 +1,249 @@
+                                  to_string(left) + ", " + to_string(right) + ").");
+  return left;
+}
+
+template <typename Backend>
+DLL_PUBLIC kernels::TensorListShape<> PropagateShapes(ExprNode &expr,
+                                                      const workspace_t<Backend> &ws) {
+  if (expr.GetNodeType() == NodeType::Constant) {
+    expr.SetShape(kernels::TensorListShape<>{{1}});
+    return expr.GetShape();
+  }
+  if (expr.GetNodeType() == NodeType::Tensor) {
+    auto &e = dynamic_cast<ExprTensor&>(expr);
+    expr.SetShape(ws.template InputRef<Backend>(e.GetMappedInput()).shape());
+    return expr.GetShape();
+  }
+  if (expr.GetSubexpressionCount() == 2) {
+    auto left_shape = PropagateShapes<Backend>(expr[0], ws);
+    auto right_shape = PropagateShapes<Backend>(expr[1], ws);
+    expr.SetShape(ShapePromotion(expr.GetOp(), left_shape, right_shape));
+    return expr.GetShape();
+  }
+  DALI_FAIL("Only binary expressions are supported");
+}
+
+template <typename Backend>
+class ArithmeticGenericOp : public Operator<Backend> {
+ public:
+  inline explicit ArithmeticGenericOp(const OpSpec &spec) : Operator<Backend>(spec) {
+    expr_ = ParseExpressionString(spec.GetArgument<std::string>("expression_desc"));
+    num_tasks_ = std::is_same<Backend, CPUBackend>::value ? 2 * num_threads_ : 1;
+  }
+
+ protected:
+  bool CanInferOutputs() const override {
+    return true;
+  }
+
+  bool SetupImpl(std::vector<OutputDesc> &output_desc, const workspace_t<Backend> &ws) override {
+    output_desc.resize(1);
+
+    result_type_id_ = PropagateTypes<Backend>(*expr_, ws);  // can be done once
+
+    result_shape_ = PropagateShapes<Backend>(*expr_, ws);
+    AllocateIntermediateNodes();
+    exec_order_ = CreateExecutionOrder<Backend>(*expr_, cache_);
+
+    output_desc[0] = {result_shape_, TypeTable::GetTypeInfo(result_type_id_)};
+    std::tie(tile_cover_, tile_range_) = GetTiledCover(result_shape_, kExtent, num_tasks_);
+    return true;
+  }
+
+  using Operator<Backend>::RunImpl;
+  void RunImpl(workspace_t<Backend> &ws) override;
+
+ private:
+  void AllocateIntermediateNodes() {
+    auto &expr = *expr_;
+    bool is_simple_expression = expr.GetSubexpressionCount() == 2 &&
+                                expr[0].GetNodeType() != NodeType::Function &&
+                                expr[1].GetNodeType() != NodeType::Function;
+    DALI_ENFORCE(is_simple_expression, "Complex expression trees are not supported");
+    // TODO(klecki): allocate memory for intermediate results and point the threads to them
+  }
+
+  std::unique_ptr<ExprNode> expr_;
+  kernels::TensorListShape<> result_shape_;
+  DALIDataType result_type_id_;
+  std::vector<TileDesc> tile_cover_;
+  std::vector<TileRange> tile_range_;
+  std::vector<ExpressionImplTask> exec_order_;
+  ExprImplCache cache_;
+  int num_tasks_;
+  static constexpr int kExtent = std::is_same<Backend, CPUBackend>::value ? 1024 : 16384;
```

Comment:
Why this values?

Reply:
Will comment, for CPU we don't won't too big intermediate buffers, for GPU we don't plan on using intermediate buffers and it's probably better to do more work at a time.

Reply:
ok